internal/app/coroutines/createCallback.go

```
@@ -1,22 +1,25 @@
 package coroutines
 
 import (
-	"fmt"
 	"log/slog"
 
 	"github.com/resonatehq/gocoro"
 	"github.com/resonatehq/resonate/internal/kernel/t_aio"
 	"github.com/resonatehq/resonate/internal/kernel/t_api"
 	"github.com/resonatehq/resonate/internal/util"
 	"github.com/resonatehq/resonate/pkg/callback"
-	"github.com/resonatehq/resonate/pkg/message"
 	"github.com/resonatehq/resonate/pkg/promise"
 )
 
 func CreateCallback(c gocoro.Coroutine[*t_aio.Submission, *t_aio.Completion, any], r *t_api.Request) (*t_api.Response, error) {
-	var res *t_api.Response
-
-	if r.CreateCallback.PromiseId == r.CreateCallback.RootPromiseId {
+	util.Assert(r.Kind == t_api.CreateCallback, "kind must be create callback")
+	util.Assert(r.CreateCallback != nil, "create callback must not be nil")
+	util.Assert(r.CreateCallback.Mesg.Type == "resume" || r.CreateCallback.Mesg.Type == "notify", "message type must be resume or notify")
+	util.Assert(r.CreateCallback.Mesg.Type == "resume" || r.CreateCallback.PromiseId == r.CreateCallback.Mesg.Root, "if notify, root promise id must equal leaf promise id")
+	// util.Assert(r.CreateCallback.Mesg.Type == "notify" ||(r.CreateCallback.PromiseId == r.CreateCallback.Mesg.Leaf && r.CreateCallback.PromiseId != r.CreateCallback.Mesg.Root), "if resume, root promise id must not equal leaf promise id")
+
+	// TODO: should this be a validation in the api layer?
```

@avillega how do you feel about moving this to the api layer?

---

I think I prefer it in the coroutine. It is kind of a business logic validation, and it applies to all api subsystems. we would have to do the validation in the http subsystem and grpc subsystem and any future ones. I prefer when all the validations are close together so it is easy to track where stuff fails.

---

I agree that it would be preferable to centralize the validation. The tricky thing is I would also prefer to "fail fast". 

Hmmm okay what about building out a validation mechanism in the kernel api implemented [here](https://github.com/resonatehq/resonate/blob/main/internal/api/api.go#L136)? I imagine we could have hooks based on the request kind, and the best part is, this is the interface that [dst uses](https://github.com/resonatehq/resonate/blob/main/test/dst/dst.go#L152) too!